EVENTS.md

```diff
@@ -0,0 +1,419 @@
+undertake, we also perform a number of further validation checks on the contracts:
+
+* No attribute key may start with `_`. This is currently used for `_contract_address` and is reserved for a 
+  namespace for injecting more *trusted* attributes from the `x/wasm` module as opposed to the contract itself
+* Event types are trimmed of whitespace, and must have at least two characters prior to prepending `wasm-`. If the contract returns
+  "  hello\n", the event type will look like `wasm-hello`. If it emits "  a  ", this will be rejected with an error (aborting execution!)
+* Attribute keys and values (both in `attributes` and under `events`) are trimmed of leading/trailing whitespace. If they are empty after
+  trimming, they are rejected as above (aborting the execution). Otherwise, they are passed verbatim.
+
+## Event Details for wasmd
+
+Beyond the basic Event system and emitted events, we must handle more advanced cases in `x/wasm`
+and thus add some more logic to the event processing. Remember that CosmWasm contracts dispatch other
+messages themselves, so far from the flattened event structure, or even a list of list (separated
+by message index in the tx), we actually have a tree of messages, each with their own events. And we must
+flatten that in a meaningful way to feed it into the event system.
+
+Furthermore, with the sub-message reply handlers, we end up with eg. "Contract A execute", "Contract B execute",
+"Contract A reply". If we return all events by all of these, we may end up with many repeated event types and
+a confusing results. However, we may use the standard "message" events to separate the sub-messages as it marks
+where the next one starts. With careful analysis of the "sender" field on these "message" markers, we may be able
+to reconstruct much of the tree execution path. We should ensure all this information is exposed in the most
+consistent way possible.
+
+### Combining Events from Sub-Messages
+
+Each time a contract is executed, it not only returns the `message` event from its call, the `execute` event for the
+contact and the `wasm` event with any custom fields from the contract itself. It will also return the same set of information
+for all messages that it returned, which were later dispatched.
+
+Until 0.18.0, we stripped out the `message` events that were returned when dispatching `Response.messages`. However,
+they contain useful information to trace the topography of the call, especially when we call into native modules,
+and it makese sense to maintain the full information.
+
+Note that these are all appended in order called. So if I execute a contract, which returns two messages, one to instantiate a new
+contract and the other to set the withdrawl address, while also using `ReplyOnSuccess` for the instantiation (to get the
+address), it will emit a series of events that looks something like this:
+
+```go
+/// original execution (top-level message)
+sdk.NewEvent(
```

Comment:
Folding in the two open issues, I would agree with filtering out all `message` events except the one top-level one (what was included in transaction), but injecting some custom start/stop events for submsg boundaries. These have no attributes so cannot be searched, only used when processing the events to recreated the execution tree.

I would propose something like this: 

```go
/// original execution (top-level message is the only one that gets the message tag)
sdk.NewEvent(
    "message",
    sdk.NewAttribute("module", "wasm"),
    sdk.NewAttribute("sender", msg.Sender),  
),

// this is where we enter the keeper, and all we would see if this was called internally from another contract
sdk.NewEvent(
    "execute",
    sdk.NewAttribute("_contract_addr", contractAddr.String()),
    sdk.NewAttribute("result", hex.EncodeToString(data)),
),
sdk.NewEvent(
    "wasm",
    sdk.NewAttribute("_contract_addr", contractAddr.String()),
    sdk.NewAttribute("custom", "from contract"),
),

// instantiating contract (returned message)
sdk.NewEvent(
    "submsg_start",
),
sdk.NewEvent(
    "instantiate",
    sdk.NewAttribute("code_id", fmt.Sprintf("%d", msg.CodeID)),
    sdk.NewAttribute("_contract_addr", newContract.String()),
    sdk.NewAttribute("result", hex.EncodeToString(initData)),
)
// didn't emit any attributes, but one event
sdk.NewEvent(
    "wasm-custom",
    sdk.NewAttribute("_contract_addr", newContract.String()),
    sdk.NewAttribute("foobar", "baz"),
),

// handling the reply
sdk.NewEvent(
    "reply",
    sdk.NewAttribute("_contract_addr", contractAddr.String()),
),
sdk.NewEvent(
    "wasm",
    sdk.NewAttribute("_contract_addr", contractAddr.String()),
    sdk.NewAttribute("custom", "from contract"),
),
// note this is emitted after handling the reply
sdk.NewEvent(
    "submsg_stop",
),

// calling the distribution module
sdk.NewEvent(
    "submsg_start",
),
sdk.NewEvent(
    "set_withdraw_address",
    sdk.NewAttribute("withdraw_address", withdrawAddr.String()),
),
sdk.NewEvent(
    "submsg_stop",
),
```

What do you think?

Reply:
Let's not add events for tracing purpose, only. reply/ start/ stop do not add value for clients IMHO

Reply:
I take your point on start and stop.

I am not sure if reply doesn't make sense. I guess it is only a separator to clarify if the calling contract did something or the subcontract did it, which may not be clear with some of the staking messages. I will think more on it.